File: apps/api/src/properties/containment.service.ts
import { HttpStatus, Injectable } from '@nestjs/common';
import { PropertiesRepository } from './properties.repository';
import { NetworkPropertyRepository } from './network-property.repository';
import { NodeScopeException } from '../common/filters/global-exception.filter';

@Injectable()
export class ContainmentService {
  constructor(
    private readonly props: PropertiesRepository,
    private readonly charters: NetworkPropertyRepository,
  ) {}

  /** PROP_007 unless `propertyId` is at/under one of `networkId`'s chartered sites. */
  async assertDevicePlacement(organizationId: string, networkId: string, propertyId: string): Promise<void> {
    const ancestors = await this.props.getAncestorIds(organizationId, propertyId);
    const charterProps = await this.charters.propertyIdsByNetwork(organizationId, networkId);
    if (!this.intersects(ancestors, charterProps)) {
      throw new NodeScopeException('PROP_007', 'DEVICE_NOT_IN_CHARTERED_SITE', HttpStatus.UNPROCESSABLE_ENTITY);
    }
  }

  /**
   * PROP_007 if moving `movedId` under `newParentId` would orphan any placed device in the moved subtree.
   * Post-move ancestors of a device d = (d's ancestors that stay within the moved subtree) ∪ (newParent + its ancestors).
   */
  async assertReparentKeepsContainment(organizationId: string, movedId: string, newParentId: string | null): Promise<void> {
    const subtree = await this.props.getSubtreeIds(organizationId, movedId);
    const devices = await this.props.devicesUnder(organizationId, subtree);
    if (devices.length === 0) return;
    const newAbove = newParentId ? await this.props.getAncestorIds(organizationId, newParentId) : [];
    const subtreeSet = new Set(subtree);
    const charterCache = new Map<string, string[]>();

    for (const d of devices) {
      const dAnc = await this.props.getAncestorIds(organizationId, d.propertyId);
      const withinMoved = dAnc.filter((a) => subtreeSet.has(a)); // d up to movedId, unchanged
      const post = new Set<string>([...withinMoved, ...newAbove]);
      let charterProps = charterCache.get(d.networkId);
      if (!charterProps) {
        charterProps = await this.charters.propertyIdsByNetwork(organizationId, d.networkId);
        charterCache.set(d.networkId, charterProps);
      }
      if (!charterProps.some((c) => post.has(c))) {
        throw new NodeScopeException('PROP_007', 'DEVICE_NOT_IN_CHARTERED_SITE', HttpStatus.UNPROCESSABLE_ENTITY);
      }
    }
  }

  /** PROP_008 if removing charter (networkId, propertyId) leaves any of that network's devices uncovered. */
  async assertCharterRemovable(organizationId: string, networkId: string, propertyId: string): Promise<void> {
    const removedSubtree = await this.props.getSubtreeIds(organizationId, propertyId);
    const devices = (await this.props.devicesUnder(organizationId, removedSubtree)).filter((d) => d.networkId === networkId);
    if (devices.length === 0) return;
    const remaining = (await this.charters.propertyIdsByNetwork(organizationId, networkId)).filter((p) => p !== propertyId);
    for (const d of devices) {
      const ancestors = await this.props.getAncestorIds(organizationId, d.propertyId);
      if (!this.intersects(ancestors, remaining)) {
        throw new NodeScopeException('PROP_008', 'CHARTER_IN_USE', HttpStatus.CONFLICT);
      }
    }
  }

  private intersects(a: string[], b: string[]): boolean {
    const set = new Set(a);
    return b.some((x) => set.has(x));
  }
}

File: src/Modules/Inventory/NodeScope.Modules.Inventory.Application/Properties/ContainmentService.cs
using NodeScope.Platform.Abstractions;

namespace NodeScope.Modules.Inventory.Application.Properties;

/// <summary>
/// The device-placement invariant (Node's <c>ContainmentService</c>): every device must sit
/// at or under one of its network's chartered sites, and no reparent or charter removal may
/// silently break that.
/// </summary>
public sealed class ContainmentService
{
    private readonly IPropertyRepository _properties;
    private readonly INetworkPropertyRepository _charters;

    public ContainmentService(IPropertyRepository properties, INetworkPropertyRepository charters)
    {
        _properties = properties;
        _charters = charters;
    }

    /// <summary><c>PROP_007</c> unless <paramref name="propertyId"/> is at/under one of the network's chartered sites.</summary>
    public async Task AssertDevicePlacementAsync(
        string organizationId,
        string networkId,
        string propertyId,
        CancellationToken cancellationToken)
    {
        var ancestors = await _properties.AncestorIdsAsync(organizationId, propertyId, cancellationToken);
        var charterProps = await _charters.PropertyIdsByNetworkAsync(organizationId, networkId, cancellationToken);
        if (!Intersects(ancestors, charterProps))
        {
            throw InventoryErrors.DeviceNotInCharteredSite();
        }
    }

    /// <summary>
    /// <c>PROP_007</c> if moving <paramref name="movedId"/> under <paramref name="newParentId"/>
    /// would orphan any placed device in the moved subtree. Post-move ancestors of a device =
    /// (its ancestors that stay within the moved subtree) union (the new parent and its ancestors).
    /// </summary>
    public async Task AssertReparentKeepsContainmentAsync(
        string organizationId,
        string movedId,
        string? newParentId,
        CancellationToken cancellationToken)
    {
        var subtree = await _properties.SubtreeIdsAsync(organizationId, movedId, cancellationToken);
        var devices = await _properties.DevicesUnderAsync(organizationId, subtree, cancellationToken);
        if (devices.Count == 0)
        {
            return;
        }

        var newAbove = newParentId is null
            ? []
            : await _properties.AncestorIdsAsync(organizationId, newParentId, cancellationToken);
        var subtreeSet = new HashSet<string>(subtree, StringComparer.Ordinal);
        var charterCache = new Dictionary<string, IReadOnlyList<string>>(StringComparer.Ordinal);

        foreach (var device in devices)
        {
            var deviceAncestors = await _properties.AncestorIdsAsync(organizationId, device.PropertyId, cancellationToken);
            var post = new HashSet<string>(deviceAncestors.Where(subtreeSet.Contains), StringComparer.Ordinal);
            post.UnionWith(newAbove);
            if (!charterCache.TryGetValue(device.NetworkId, out var charterProps))
            {
                charterProps = await _charters.PropertyIdsByNetworkAsync(organizationId, device.NetworkId, cancellationToken);
                charterCache[device.NetworkId] = charterProps;
            }

            if (!charterProps.Any(post.Contains))
            {
                throw InventoryErrors.DeviceNotInCharteredSite();
            }
        }
    }

    /// <summary><c>PROP_008</c> if removing the charter leaves any of the network's devices uncovered.</summary>
    public async Task AssertCharterRemovableAsync(
        string organizationId,
        string networkId,
        string propertyId,
        CancellationToken cancellationToken)
    {
        var removedSubtree = await _properties.SubtreeIdsAsync(organizationId, propertyId, cancellationToken);
        var devices = (await _properties.DevicesUnderAsync(organizationId, removedSubtree, cancellationToken))
            .Where(device => device.NetworkId == networkId)
            .ToList();
        if (devices.Count == 0)
        {
            return;
        }

        var remaining = (await _charters.PropertyIdsByNetworkAsync(organizationId, networkId, cancellationToken))
            .Where(id => id != propertyId)
            .ToList();
        foreach (var device in devices)
        {
            var ancestors = await _properties.AncestorIdsAsync(organizationId, device.PropertyId, cancellationToken);
            if (!Intersects(ancestors, remaining))
            {
                throw InventoryErrors.CharterInUse();
            }
        }
    }

    private static bool Intersects(IReadOnlyCollection<string> a, IReadOnlyCollection<string> b)
    {
        var set = new HashSet<string>(a, StringComparer.Ordinal);
        return b.Any(set.Contains);
    }
}
